util/noeintr_close.c

```diff
@@ -0,0 +1,284 @@
+ * and still open" by seeing if MSG_PEEK returns our magic cookie.
+ *
+ * If we're on a perverse OS where close(fd) can successfully close the
+ * descriptor and still return -1 / EINTR, *and* another thread wins a race
+ * and reuses that descriptor, there's still no way we'll see our random
+ * cookie, so we won't make the mistake of trying to close it again.
+ */
+#include <sys/socket.h>
+
+#include <errno.h>
+#include <pthread.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "entropy.h"
+#include "optional_mutex.h"
+#include "warnp.h"
+
+#include "noeintr.h"
+
+#ifdef NOEINTR_CLOSE_TESTING
+#include "noeintr_close_testing.h"
+#endif
+
+/* Synchronization. */
+static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+
+/* Initialization. */
+static enum noeintr_close_status {
+	NOEINTR_UNINITIALIZED,
+	NOEINTR_INITIALIZED,
+	NOEINTR_FATAL
+} status = NOEINTR_UNINITIALIZED;
+
+/* Magic cookie. */
+#define MAGIC_COOKIE_LEN 32
+static int s[2] = {-1, -1};
+static uint8_t cookie_write[MAGIC_COOKIE_LEN];
+static struct entropy_read_cookie * entropy_read_cookie = NULL;
+
+/* Clean up on program exit. */
+static void
+noeintr_close_atexit(void)
+{
+
+	/*
+	 * Clean up the entropy_read_cookie; this may call noeintr_close().
+	 * As such, it must be handled before freeing any other resources.
+	 */
+	if (entropy_read_cookie != NULL)
+		entropy_read_done(entropy_read_cookie);
+
+	/* Attempt to close s[0] and s[1] using the normal close(). */
+	if ((s[0] != -1) && close(s[0]))
+		warnp("close");
+	if ((s[1] != -1) && close(s[1]))
+		warnp("close");
+}
+
+/* Initialize socketpair, atexit, and magic cookie. */
+static int
+noeintr_close_init(void)
+{
+
+	/* Create a socket pair. */
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, s)) {
+		warnp("socketpair");
+		goto err1;
+	}
+
+	/* Set up cleanup function. */
+	if (atexit(noeintr_close_atexit)) {
+		warnp("atexit");
+		goto err2;
+	}
+
+	/* Entropy cookie; don't clean it until noeintr_close_atexit(). */
+	if ((entropy_read_cookie = entropy_read_init()) == NULL) {
+		warn0("entropy_read_init");
+		goto err2;
+	}
+
+	/* Generate a random cookie. */
+	if (entropy_read_fill(entropy_read_cookie, cookie_write,
+	    MAGIC_COOKIE_LEN)) {
+		warn0("entropy_read");
+		goto err2;
+	}
+
+	/* Write the cookie to s[0]. */
+	if (noeintr_write(s[0], cookie_write, MAGIC_COOKIE_LEN)
+	    != MAGIC_COOKIE_LEN) {
+		warnp("noeintr_write");
+		goto err2;
+	}
+
+	/* We've finished initializing. */
+	status = NOEINTR_INITIALIZED;
+
+	/* Success! */
+	return (0);
+
+err2:
+	/* Attempt to close s[0] and s[1] using the normal close(). */
+	if (close(s[0]))
+		warnp("close");
+	s[0] = -1;
+	if (close(s[1]))
+		warnp("close");
+	s[1] = -1;
+err1:
+	status = NOEINTR_FATAL;
+
+	/* Failure! */
+	return (-1);
+}
+
+/* Check if ${fd} produces our magic cookie.  Return 0 if it doesn't. */
+static int
+is_fd_still_open(int fd)
```

Comment:
I don't like this function name (and I'm certain that you'll hate it), but I definitely think it's worth doing this in a separate function, rather than as part of `noeintr_close()` itself.